### PR TITLE
Simple logging using the standard library

### LINKED ISSUE #930: Improve logging
Goals:

- [x] No interleaved printing in a multi-threaded environment
- [x] Associate a log line with the caller context which is the thread, the file, and its line 
- [x] Configurable, so we can have more or less noise in the console. The default for our current `debug: true` would probably be 'INFO', for `debug: false` it would still be either 'ERROR' or 'WARNING'
- [x] Maybe a hook ('Handler') that wires up 'ERROR' messages with a toast plugin or an (yet to come) error panel, or a just status bar message to keep it simple 🌶 
- [x] If the user says `debug: false` (the current default) we do not format the messages fancy with thread, filename, line; we just print the actual *message* 🎆 
- [x] A plugin author justs import the std python *logging* library and inits it to e.g. `getLogger('SublimeLinter.flake8')` and then logs away, and it will cooperate with our Logger. T.i. the author will not use `persist.debug` anymore ❤️ 

Related #909

### PR DESCRIPTION
**Pull request by kaste**
Fixes #930

### CODE REVIEW

**Comment by braver** [comment]
It looks good basically, but it seems "info" level logs don't get printed when I set debug to true. For the end user I expect there to be just an on-and-off switch and "info" has important information right now.

**Comment by kaste** [comment]
? This is what I get with `debug: true`

image

**Comment by braver** [comment]
Specifically I was missing the version info that’s printed on start up.

**Comment by kaste** [comment]
Good catch. This is random behavior actually. 

I thought naming the file `_logging.py` would ensure that its `plugin_loaded()` callback would be invoked before any other thing. But that's not the case, loading order here is random.

**Comment by kaste** [comment]
I'm not so much into log message styling. So this LGTM.

**Comment by braver** [comment]
👍

**Comment by braver** [comment]
#909 can be closed then I think?

**Comment by kaste** [comment]
Current status: 12 issues and 0 PRs.

**Comment by kaste** [comment]
image

**Comment by braver** [comment]
🎉 Let's freeze it, check some docs and then release it. I'm away-from-laptop from Thursday till the end of next week, so let's pin release day on Feb 24?

**Comment by braver** [comment]
Just in time too, I just tagged the penultimate pre-release @ `4.0.0-rc.9.9.8`. After `4.0.0-rc.9.9.8` the numbering becomes super ridiculous.

File: lint/linter.py [hunk 1]
```diff
@@ -921,9 +925,19 @@ def lint(self, code, hit_time, settings):
         - If the view has been modified since the original hit_time, stop.
         - Parse the linter output with the regex.
         """
+        # We 'name' our threads, for logging purposes.
+        threading.current_thread().name = task_name
```
Comment (FichteFoll) [inline lint/linter.py]: I believe this should be happening in `execute_lint_task` instead. `Linter.lint` does not need to know or care about the backend being used and how tasks are scheduled.

File: lint/log_handler.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        formatter = TaskNumberFormatter(
+            fmt="SublimeLinter: {LEVELNAME}{message}",
+            style='{')
+        handler = logging.StreamHandler()
+        handler.setFormatter(formatter)
+    else:
+        if level is True:
+            level = DEBUG_TRUE_LEVEL
+        else:
+            level = logging.getLevelName(level.upper())
```
Comment (FichteFoll) [inline lint/log_handler.py]: Shouldn't this be `getattr(logging, level.upper())`? `logging.getLevelName` is the reverse operation.

File: lint/linter.py (hunk 1 above)
Comment (kaste) [reply lint/linter.py]: Yeah, that would be better

File: lint/log_handler.py (hunk 2 above)
Comment (kaste) [reply lint/log_handler.py]: Hm, I have it from here https://github.com/python/cpython/blob/504f19145ca5738162d6a720fa45b364ac8c0384/Lib/logging/__init__.py#L120-L141

Comment (FichteFoll) [reply lint/log_handler.py]: Yeah, but you already have a string in `level` and want the numeric level for `handler.setLevel`.

Comment (kaste) [reply lint/log_handler.py]: You did not read the code, did you

Comment (FichteFoll) [reply lint/log_handler.py]: No, I actually did not. I only read the docstring and it looks like this behavior is undocumented (or rather a backwards compatibility thing that has been removed for 3.4) which is why this behavior confused me. And admittedly, for a function named `getLevelName` to return an integer with the level (and not its name) *is* pretty confusing.

Related upstream issue: https://bugs.python.org/issue22386

**Pull request merged**